posthog/migrations/0676_add_question_ids_to_surveys.py

```diff
@@ -0,0 +1,45 @@
+# Generated by Django 4.2.18 on 2025-02-25 19:29
+
+import uuid
+from django.db import migrations
+
+
+def add_question_ids_to_surveys(apps, schema_editor):
+    """
+    Add a unique ID to each question in existing surveys.
+    This ensures all questions have an ID for tracking responses.
+    """
+    Survey = apps.get_model("posthog", "Survey")
+
+    # Get all surveys
+    surveys = Survey.objects.all()
+
+    for survey in surveys:
+        # Skip surveys with no questions
+        if not survey.questions:
+            continue
+
+        # Add IDs to questions that don't have them
+        modified = False
+        for question in survey.questions:
+            if not question.get("id"):
+                question["id"] = str(uuid.uuid4())
+                modified = True
+
+        # Save the survey if any questions were modified
+        if modified:
+            survey.save(update_fields=["questions"])
```

Comment:
2 questions:
1. `Survey.objects.all()` can be quite heavy, is it ok to do this? can we check with either CDP or infra folks? not sure if this would be fast or not since all the JSON serialization.
2. this block isn't synchronized, so if a survey is edited within the middle of the process, maybe something can be messed up?

Reply:
running with a transaction is different than being a synchronized block.
my question is if surveys are being edited/created while the migration is running, will that be ok?

Reply:
yes.

1. the migration's view of data and an user concurrent edits/adds are isolated until their are commited
2. there's no schema changes, we're only modifying data. which means we won't have any locking on the DB
3. the operation is idempotent. no matter how many times it's run, it'll only add ID for questions once. in the worst case, a question might get a UUID from the user instead of from the migration, but it doesn't matter, as long as it gets one UUID.

if a user loads a survey for editing before the migration is processed, then, they can potentially save it after the migration has already modified them (from the `ensure_question_ids` pre_save hook). but, we still check if the question.id is not there, and only then we update it

Reply:
`surveys = Survey.objects.all()` here you load all surveys in memory, all of them don't have IDs
meanwhile traversing the questions and saving everything within a transaction might take a few minutes (I guess?).
if a survey is edited and new IDs are added with the `pre_save_survey`, now the code here will replace all IDs for that survey since when you loaded in memory, they did not have IDs.
This is a concurrency issue since you are loading everything in memory so it allows time for this to happen.
how can we prevent this? should we prevent this?

Reply:
deleted the migration; now running it as a script.

on the script:

1. no longer loading all surveys on memory; instead, doing on a batch of 1k
2. when we get the surveys, we actually lock them to prevent updates on them. this way, no IDs will be overwritten. and, if it finds any row already locked, it'll re-run those after